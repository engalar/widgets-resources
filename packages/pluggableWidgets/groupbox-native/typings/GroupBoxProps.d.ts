/**
 * This file was generated from GroupBox.xml
 * WARNING: All changes made to this file will be overwritten
 * @author Mendix UI Content Team
 */
import { ComponentType, ReactNode } from "react";
import { DynamicValue, NativeIcon } from "mendix";

export type CollapsibleEnum = "collapsibleNo" | "collapsibleYesExpanded" | "collapsibleYesCollapsed";

export interface GroupBoxProps<Style> {
    name: string;
    style: Style[];
    headerContent?: ReactNode;
    content: ReactNode;
    showHeader: boolean;
    collapsible: CollapsibleEnum;
    iconCollapsed?: DynamicValue<NativeIcon>;
    iconExpanded?: DynamicValue<NativeIcon>;
}

export interface GroupBoxPreviewProps {
    class: string;
    style: string;
    headerContent: { widgetCount: number; renderer: ComponentType<{caption?: string}> };
    content: { widgetCount: number; renderer: ComponentType<{caption?: string}> };
    showHeader: boolean;
    collapsible: CollapsibleEnum;
    iconCollapsed: { type: "glyph"; iconClass: string; } | { type: "image"; imageUrl: string; } | null;
    iconExpanded: { type: "glyph"; iconClass: string; } | { type: "image"; imageUrl: string; } | null;
}
