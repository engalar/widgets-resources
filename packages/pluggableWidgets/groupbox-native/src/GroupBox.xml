<?xml version="1.0" encoding="utf-8" ?>
<widget id="com.mendix.widget.native.groupbox.GroupBox" supportedPlatform="Native" offlineCapable="true" pluginWidget="true" xmlns="http://www.mendix.com/widget/1.0/" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://www.mendix.com/widget/1.0/ ../../../node_modules/mendix/custom_widget.xsd">
    <name>Group Box</name>
    <description>Group content in a container.</description>
    <helpUrl>https://github.com/mendix/widgets-resources/blob/master/packages/pluggableWidgets/groupbox-native/README.md</helpUrl>
    <icon>iVBORw0KGgoAAAANSUhEUgAAAEAAAABACAYAAACqaXHeAAAACXBIWXMAAAsTAAALEwEAmpwYAAAAAXNSR0IArs4c6QAAAARnQU1BAACxjwv8YQUAAAFuSURBVHgB7ds/UoNAGAXwR4wN2CZ2tmk8BIW5AdBoQeeVLGHGCq5gwSFsbO0cWqGJTgRnLJQ/O+uQAt77lZsvM9k3fLubzMY5NkBsBXIKAOTWfweKokCSJKjrGkuy2WwQhiF83/813nkC8jxf3ORbZVkiTdPO+KqvcKmqquqMaQ0wFWRZhjmLomj0dW2DIKc1AJaeXj/x8PyB6jDtV4hL18Hdbo2bq7PRuqFzytA+b2L9BDy+TD/51lt9bII9GOuGzilD+7yJdQDv5s/4bxfnjrGmby//4XkebFkHcH+9/n5cp+Y1k7/dmTsyjmNst9vu+5vJB0EAW9ZrwL7p0b2hT0+p7XHbPh+jbRDkdA4wFZjO0nOnFgA5rQGmAv0esHAKAOQUAMgpAJBTACCnAEBOAYCcAgA5BQByCgDkFADIKQCQUwAgpwBATgGAnAIAOd0RArlOAK7rYqn6rth2Ahi6izt3Q3eJHf13mJwCALkvY/Jc3wh3q30AAAAASUVORK5CYII=</icon>
    <properties>
        <propertyGroup caption="General">
            <propertyGroup caption="General">
                <property key="headerContent" type="widgets" required="false">
                    <caption>Header</caption>
                    <description />
                </property>
                <property key="content" type="widgets" required="true">
                    <caption>Content</caption>
                    <description />
                </property>
                <property key="showHeader" type="boolean" defaultValue="true">
                    <caption>Show header</caption>
                    <description></description>
                </property>
                <property key="collapsible" type="enumeration" defaultValue="collapsibleYesExpanded">
                    <caption>Collapsible</caption>
                    <description></description>
                    <enumerationValues>
                        <enumerationValue key="collapsibleNo">No</enumerationValue>
                        <enumerationValue key="collapsibleYesExpanded">Yes (start expanded)</enumerationValue>
                        <enumerationValue key="collapsibleYesCollapsed">Yes (start collapsed)</enumerationValue>
                    </enumerationValues>
                </property>
                <property key="iconCollapsed" type="icon" required="false">
                    <caption>Icon collapsed</caption>
                    <description/>
                </property>
                <property key="iconExpanded" type="icon" required="false">
                    <caption>Icon expanded</caption>
                    <description/>
                </property>
            </propertyGroup>
            <propertyGroup caption="Common">
                <systemProperty key="Name" />
                <systemProperty key="Visibility" />
            </propertyGroup>
        </propertyGroup>
    </properties>
</widget>
