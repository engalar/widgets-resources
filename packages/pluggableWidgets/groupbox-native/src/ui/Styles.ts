import { ViewStyle } from "react-native";
import { Style } from "@mendix/piw-native-utils-internal";

export interface GroupBoxStyle extends Style {
    container: ViewStyle;
    header: {
        container: ViewStyle;
        content: ViewStyle;
        icon: ViewStyle & {
            size: number;
            color: string;
        };
    };
    content: ViewStyle;
}

export const defaultGroupBoxStyle: GroupBoxStyle = {
    container: {
            borderWidth: 1,
            borderColor: "#e7e7e9",
            borderRadius: 4,
            overflow: "hidden"
    },
    header: {
        container: {
            flex: 1,
            flexDirection: "row",
            alignItems: "center",
            backgroundColor: "#e7e7e9",
            paddingHorizontal: 12,
            paddingVertical: 8,
        },
        content: {
            flex: 1,
        },
        icon: {
            size: 16,
            color: "#0a1325",
        }
    },
    content: {
        paddingVertical: 12,
        paddingHorizontal: 8,
    }
};
