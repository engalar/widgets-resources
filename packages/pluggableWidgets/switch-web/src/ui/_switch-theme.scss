$default-android-color: #6fbeb5;
$default-ios-color: rgb(100, 189, 99);

@mixin bootstrap-style-ios($brand-style) {
    background-color: $brand-style;
    border-color: $brand-style;
    box-shadow: $brand-style 0 0 0 16px inset;
}

@mixin bootstrap-style-android($brand-style) {
    background-color: lighten($brand-style, 10%);
}

.widget-switch {
    &.iOS,
    .auto {
        .widget-switch-btn-wrapper {
            &.checked {
                &.widget-switch-btn-wrapper-default {
                    @include bootstrap-style-ios($default-ios-color);
                }
            }
        }
    }

    &.android {
        .widget-switch-btn-wrapper {
            &.checked {
                &.widget-switch-btn-wrapper-default {
                    @include bootstrap-style-android($default-android-color);

                    .widget-switch-btn {
                        background: $default-android-color;
                    }
                }
            }
        }
    }
}
