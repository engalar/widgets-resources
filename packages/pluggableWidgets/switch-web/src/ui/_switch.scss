@import "mixins";

.widget-switch-btn-wrapper {
    margin-bottom: 8px;

    &:focus {
        outline: 1px solid #0595db;
    }

    &.disabled {
        cursor: default;
    }
}

.widget-switch {
    .widget-switch-checkbox {
        display: none;
    }

    &.iOS {
        @include ios;
    }

    &.android {
        @include android;
    }

    &.auto {
        @include ios;
    }
}

.mx-table {
    .widget-switch {
        &.android {
            padding-top: 6px;
            padding-bottom: 7px;
        }
    }
}
