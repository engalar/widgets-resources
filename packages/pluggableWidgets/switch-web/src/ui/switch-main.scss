@import "Switch";
@import "switch-theme";
