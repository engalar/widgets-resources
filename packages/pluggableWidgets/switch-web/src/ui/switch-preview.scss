@import "mixins";
@import "variables";

$brand-primary: #264ae5;
$brand-success: #3cb33d;
$brand-warning: #eca51c;
$brand-danger: #e33f4e;

@import "switch";

@include style("iOS", "default", $default-ios-color);
@include style("iOS", "primary", $brand-primary);
@include style("iOS", "success", $brand-success);
@include style("iOS", "warning", $brand-warning);
@include style("iOS", "danger", $brand-danger);

@include style("android", "default", $default-android-color);
@include style("android", "primary", $brand-primary);
@include style("android", "success", $brand-success);
@include style("android", "warning", $brand-warning);
@include style("android", "danger", $brand-danger);
