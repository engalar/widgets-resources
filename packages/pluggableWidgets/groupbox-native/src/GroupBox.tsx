import { createElement, ReactElement, ReactNode, useCallback, useEffect, useRef, useState } from "react";
import { Animated, Easing, LayoutChangeEvent, Pressable, View, ViewStyle } from "react-native";
import { Icon } from "mendix/components/native/Icon";
import { exclude, flattenStyles } from "@mendix/piw-native-utils-internal";

import { GroupBoxStyle, defaultGroupBoxStyle } from "./ui/Styles";
import { GroupBoxProps } from "../typings/GroupBoxProps";

declare interface GlyphIcon {
    readonly type: "glyph";
    readonly iconClass: string;
}

export function GroupBox(props: GroupBoxProps<GroupBoxStyle>): ReactElement | null {
    const styles = flattenStyles(defaultGroupBoxStyle, props.style);
    const startedExpanded = !props.showHeader || props.collapsible === "collapsibleYesExpanded";
    const [isExpanded, setIsExpanded] = useState(startedExpanded);

    const toggleContent = () => {
        setIsExpanded(value => !value);
    };

    const renderIcon = () => {
        const { iconCollapsed, iconExpanded } = props;
        const customIconsConfigured = iconCollapsed && iconCollapsed.value && iconExpanded && iconExpanded.value;
        const customIconSource = iconCollapsed?.value;
        const customExpandedIconSource = iconExpanded?.value;
        const source = isExpanded ? customExpandedIconSource : customIconSource;
        const iconStyles = exclude(styles.header.icon, ["size", "color"]);

        return customIconsConfigured ? (
            <View style={iconStyles}>
                <Icon icon={source} size={styles.header.icon.size} color={styles.header.icon.color} />
            </View>
        ) : (
            <AnimatedRotatingIconChevron
                isExpanded={isExpanded}
                style={iconStyles}
                size={styles.header.icon.size}
                color={styles.header.icon.color}
            />
        );
    };
    return (
        <View style={styles.container}>
            {props.showHeader && (
                <Pressable
                    testID={"header"}
                    style={styles.header.container}
                    onPress={props.collapsible !== "collapsibleNo" ? toggleContent : null}
                >
                    <View style={styles.header.content}>{props.headerContent}</View>
                    {props.collapsible !== "collapsibleNo" && renderIcon()}
                </Pressable>
            )}
            <AnimatedCollapsibleView isExpanded={isExpanded} style={styles.content}>
                {props.content}
            </AnimatedCollapsibleView>
        </View>
    );
}

export function AnimatedRotatingIconChevron({
    isExpanded,
    style,
    size,
    color
}: {
    isExpanded: boolean;
    style: ViewStyle;
    size: number;
    color: string;
}) {
    const icon: GlyphIcon = { type: "glyph", iconClass: "glyphicon-chevron-down" };
    const animatedValue = useRef(new Animated.Value(0)).current;
    const animatedRotation = animatedValue.interpolate({
        inputRange: [0, 1],
        outputRange: ["0deg", "180deg"]
    });

    useEffect(() => {
        Animated.timing(animatedValue, {
            toValue: isExpanded ? 1 : 0,
            duration: 200,
            easing: Easing.ease,
            useNativeDriver: false
        }).start();
    }, [isExpanded, animatedValue]);

    return (
        <Animated.View style={[style, { height: size, width: size, transform: [{ rotate: animatedRotation }] }]}>
            <Icon icon={icon} size={size} color={color} />
        </Animated.View>
    );
}

export function AnimatedCollapsibleView({
    isExpanded,
    style,
    children
}: {
    isExpanded: boolean;
    style: ViewStyle;
    children: ReactNode;
}): ReactElement {
    const startingHeight = 0;
    const animatedHeight = useRef(new Animated.Value(startingHeight)).current;
    const [fullHeight, setFullHeight] = useState(startingHeight);
    const isFullHeightCalculated = useRef(false);

    useEffect(() => {
        Animated.timing(animatedHeight, {
            toValue: isExpanded ? fullHeight : startingHeight,
            duration: 300,
            easing: Easing.ease,
            useNativeDriver: false
        }).start();
    }, [isExpanded, fullHeight, animatedHeight]);

    const onLayout = useCallback(
        (e: LayoutChangeEvent) => {
            if (!isFullHeightCalculated.current) {
                isFullHeightCalculated.current = true;
                setFullHeight(e.nativeEvent.layout.height);
            }
        },
        [isFullHeightCalculated.current]
    );

    return (
        <Animated.View
            style={{
                height: isFullHeightCalculated.current ? animatedHeight : undefined
            }}
            onLayout={onLayout}
        >
            <View style={style}>{children}</View>
        </Animated.View>
    );
}
