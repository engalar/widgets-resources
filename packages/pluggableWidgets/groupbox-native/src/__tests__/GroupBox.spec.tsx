import { createElement } from "react";
import { Pressable, Text } from "react-native";
import { fireEvent, render } from "react-native-testing-library";
import { dynamicValue } from "@mendix/piw-utils-internal";

import { GroupBox } from "../GroupBox";
import { GroupBoxProps } from "../../typings/GroupBoxProps";
import { GroupBoxStyle } from "../ui/Styles";

const defaultProps: GroupBoxProps<GroupBoxStyle> = {
    name: "groupBoxTest",
    style: [],
    showHeader: true,
    collapsible: "collapsibleYesExpanded",
    headerContent: <Text>Header</Text>,
    content: <Text>Content</Text>,
    iconCollapsed: dynamicValue({ type: "glyph", iconClass: "glyphicon-chevron-down" }),
    iconExpanded: dynamicValue({ type: "glyph", iconClass: "glyphicon-chevron-up" })
};

describe("GroupBox", () => {
    beforeEach(() => {
        jest.useFakeTimers();
    });

    it("renders with default styles", () => {
        const component = render(<GroupBox {...defaultProps} />);

        expect(component.toJSON()).toMatchSnapshot();
    });

    it("renders with custom styles", () => {
        const style: GroupBoxStyle[] = [
            {
                container: {},
                header: {
                    container: {},
                    content: {},
                    icon: {
                        size: 16,
                        color: "#000"
                    }
                },
                content: {}
            }
        ];

        const component = render(<GroupBox {...defaultProps} style={style} />);

        expect(component.toJSON()).toMatchSnapshot();
    });

    it("renders without content", () => {
        const component = render(<GroupBox {...defaultProps} content={null} />);

        expect(component.toJSON()).toMatchSnapshot();
    });

    it("renders without header", () => {
        const component = render(<GroupBox {...defaultProps} showHeader={false} />);

        expect(component.toJSON()).toMatchSnapshot();
    });

    it("renders without header content", () => {
        const component = render(<GroupBox {...defaultProps} headerContent={null} />);

        expect(component.toJSON()).toMatchSnapshot();
    });

    it("renders without being collapsible", () => {
        const component = render(<GroupBox {...defaultProps} collapsible={"collapsibleNo"} />);
        component.findByTestId("header").then(header => {
            const json = component.toJSON();
            expect(json).toMatchSnapshot();
            fireEvent.press(header);
            expect(json).toMatchSnapshot();
        });
    });

    it("expands content when header is clicked", () => {
        const component = render(<GroupBox {...defaultProps} collapsible={"collapsibleYesCollapsed"} />);
        component.findByTestId("header").then(header => {
            const json = component.toJSON();
            expect(json).toMatchSnapshot();
            fireEvent.press(header);
            expect(json).not.toMatchSnapshot();
        });
    });

    it("expands content when header is empty and is clicked", () => {
        const component = render(<GroupBox {...defaultProps} headerContent={null} collapsible={"collapsibleYesCollapsed"} />);
        component.findByTestId("header").then(header => {
            const json = component.toJSON();
            expect(json).toMatchSnapshot();
            fireEvent.press(header);
            expect(json).not.toMatchSnapshot();
        });
    });

    it("does not expand content when a clickable component in the header is clicked", () => {
        const headerContent = <Pressable testID={"customHeaderButton"}>Button</Pressable>;
        const component = render(<GroupBox {...defaultProps} headerContent={headerContent} />);
        component.findByTestId("customHeaderButton").then(button => {
            const json = component.toJSON();
            expect(json).toMatchSnapshot();
            fireEvent.press(button);
            expect(json).toMatchSnapshot();
        });
    });
});
