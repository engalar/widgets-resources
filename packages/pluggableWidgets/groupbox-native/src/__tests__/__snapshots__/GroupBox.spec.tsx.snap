// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GroupBox renders with custom styles 1`] = `
<View
  style={
    Object {
      "borderColor": "#e7e7e9",
      "borderRadius": 4,
      "borderWidth": 1,
      "overflow": "hidden",
    }
  }
>
  <View
    accessible={true}
    focusable={true}
    onBlur={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={Object {}}
    testID="header"
  >
    <View
      style={Object {}}
    >
      <Text>
        Header
      </Text>
    </View>
    <View
      style={Object {}}
    >
      <View
        testId="icon"
      />
    </View>
  </View>
  <View
    onLayout={[Function]}
    style={
      Object {
        "height": undefined,
      }
    }
  >
    <View
      style={
        Object {
          "paddingHorizontal": 8,
          "paddingVertical": 12,
        }
      }
    >
      <Text>
        Content
      </Text>
    </View>
  </View>
</View>
`;

exports[`GroupBox renders with default styles 1`] = `
<View
  style={
    Object {
      "borderColor": "#e7e7e9",
      "borderRadius": 4,
      "borderWidth": 1,
      "overflow": "hidden",
    }
  }
>
  <View
    accessible={true}
    focusable={true}
    onBlur={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "#e7e7e9",
        "flex": 1,
        "flexDirection": "row",
        "paddingHorizontal": 12,
        "paddingVertical": 8,
      }
    }
    testID="header"
  >
    <View
      style={
        Object {
          "flex": 1,
        }
      }
    >
      <Text>
        Header
      </Text>
    </View>
    <View
      style={Object {}}
    >
      <View
        testId="icon"
      />
    </View>
  </View>
  <View
    onLayout={[Function]}
    style={
      Object {
        "height": undefined,
      }
    }
  >
    <View
      style={
        Object {
          "paddingHorizontal": 8,
          "paddingVertical": 12,
        }
      }
    >
      <Text>
        Content
      </Text>
    </View>
  </View>
</View>
`;

exports[`GroupBox renders without content 1`] = `
<View
  style={
    Object {
      "borderColor": "#e7e7e9",
      "borderRadius": 4,
      "borderWidth": 1,
      "overflow": "hidden",
    }
  }
>
  <View
    accessible={true}
    focusable={true}
    onBlur={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "#e7e7e9",
        "flex": 1,
        "flexDirection": "row",
        "paddingHorizontal": 12,
        "paddingVertical": 8,
      }
    }
    testID="header"
  >
    <View
      style={
        Object {
          "flex": 1,
        }
      }
    >
      <Text>
        Header
      </Text>
    </View>
    <View
      style={Object {}}
    >
      <View
        testId="icon"
      />
    </View>
  </View>
  <View
    onLayout={[Function]}
    style={
      Object {
        "height": undefined,
      }
    }
  >
    <View
      style={
        Object {
          "paddingHorizontal": 8,
          "paddingVertical": 12,
        }
      }
    />
  </View>
</View>
`;

exports[`GroupBox renders without header 1`] = `
<View
  style={
    Object {
      "borderColor": "#e7e7e9",
      "borderRadius": 4,
      "borderWidth": 1,
      "overflow": "hidden",
    }
  }
>
  <View
    onLayout={[Function]}
    style={
      Object {
        "height": undefined,
      }
    }
  >
    <View
      style={
        Object {
          "paddingHorizontal": 8,
          "paddingVertical": 12,
        }
      }
    >
      <Text>
        Content
      </Text>
    </View>
  </View>
</View>
`;

exports[`GroupBox renders without header content 1`] = `
<View
  style={
    Object {
      "borderColor": "#e7e7e9",
      "borderRadius": 4,
      "borderWidth": 1,
      "overflow": "hidden",
    }
  }
>
  <View
    accessible={true}
    focusable={true}
    onBlur={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "#e7e7e9",
        "flex": 1,
        "flexDirection": "row",
        "paddingHorizontal": 12,
        "paddingVertical": 8,
      }
    }
    testID="header"
  >
    <View
      style={
        Object {
          "flex": 1,
        }
      }
    />
    <View
      style={Object {}}
    >
      <View
        testId="icon"
      />
    </View>
  </View>
  <View
    onLayout={[Function]}
    style={
      Object {
        "height": undefined,
      }
    }
  >
    <View
      style={
        Object {
          "paddingHorizontal": 8,
          "paddingVertical": 12,
        }
      }
    >
      <Text>
        Content
      </Text>
    </View>
  </View>
</View>
`;
